[Docs] Document Encrypted Saved Objects functionality. (#80183)
elastic/kibana: docs/user/security/secure-saved-objects.asciidoc

Example `kibana.yml`:

[IMPORTANT]
============================================================================
If you don't specify encryption key {kib} will generate a random one automatically at startup time. That means every time you restart {kib} it will use a new ephemeral encryption key and won't be able to decrypt saved objects encrypted with another key. To prevent data loss, {kib} may disable features which rely on this encryption until you explicitly set an encryption key.
============================================================================

[source,yaml]
--------------------------------------------------------------------------------
xpack.encryptedSavedObjects:
  encryptionKey: "min-32-byte-long-strong-encryption-key"
--------------------------------------------------------------------------------

[[encryption-key-rotation]]
==== Encryption key rotation

Many policies and best practices stipulate that encryption keys should be periodically rotated to decrease the amount of content encrypted with one key and hence limit the potential damage in case the key is compromised. {kib} allows you to rotate encryption keys whenever there is a need. 

> As a rule of thumb, when you change an encryption key you still keep the old one for some time. While {kib} will only use a new encryption key to encrypt all new and updated data it still may need to know the old one to decrypt data that was encrypted using that old key. Moreover, it's perfectly possible to have multiple old keys used only for decryption. It's worth noting that {kib} doesn't re-encrypt existing saved objects with the new encryption key automatically, re-encryption only happens when you update existing object or use <<saved-objects-api-rotate-encryption-key, Rotate encryption key API>>.

Reviewer: ```suggestion
When you change an encryption key, keep the old one for some time. Although {kib} only uses a new encryption key to encrypt all new and updated data, it still may need the old one to decrypt data that was encrypted using the old key. It's possible to have multiple old keys used only for decryption. {kib} doesn't automatically re-encrypt existing saved objects with the new encryption key. Re-encryption only happens when you update existing object or use the <<saved-objects-api-rotate-encryption-key, rotate encryption key API>>.
```

Author: `As a rule of thumb, when you change an encryption key you still keep the old one for some time.` ---> `When you change an encryption key, keep the old one for some time. ` this one is a bit tricky, we don't want to force/recommend and be responsible for that decision, we just say it's usually done like this. But there could be cases when encryption key leaks, data encrypted with it was tampered with something malicious and users really don't want these objects to be decrypted and used for whatever purpose in production. That means old keys should just be disposed immediately. We cannot also call this out specifically since their Security team should know what would be appropriate in their particular case.

@gchaps can you please suggest the appropriate language/tone for this?

Reviewer: This makes the sentence stronger:

When you change an encryption key, be sure to keep the old one for some time.

> Here is how your `kibana.yml` can look like if you want to use key rotation functionality:

Reviewer: ```suggestion
Here is how your `kibana.yml` might look if you use key rotation functionality:
```

[source,yaml]
--------------------------------------------------------------------------------
xpack.encryptedSavedObjects:
  encryptionKey: "min-32-byte-long-NEW-encryption-key" <1>
  keyRotation:
    decryptionOnlyKeys: ["min-32-byte-long-OLD#1-encryption-key", "min-32-byte-long-OLD#2-encryption-key"] <2>
--------------------------------------------------------------------------------

> <1> Encryption key {kib} will use to encrypt all new or updated saved objects, known as the primary encryption key.

Reviewer: ```suggestion
<1> The encryption key {kib} will use to encrypt all new or updated saved objects.  This is known as the primary encryption key.
```

> <2> A list of encryption keys {kib} will try to use only to decrypt existing saved objects if decryption with the primary encryption key isn't possible, known as decryption-only or secondary encryption keys.

Reviewer: ```suggestion
<2> A list of encryption keys {kib} will try to use to decrypt existing saved objects if decryption with the primary encryption key isn't possible.  These keys are known as the decryption-only or secondary encryption keys.
```

[NOTE]
============================================================================

> You may also leverage this functionality if, by mistake, you ended up with multiple {kib} instances connected to the same {es} cluster that use different encryption keys. In this case you may have a mix of saved objects encrypted with different keys and every {kib} instance can deal only with a specific subset of objects. To fix this you'll need to choose a single primary encryption key for `xpack.encryptedSavedObjects.encryptionKey`, move all other encryption keys to `xpack.encryptedSavedObjects.keyRotation.decryptionOnlyKeys` and sync this configuration across all {kib} instances.

Reviewer: ```suggestion
You might also leverage this functionality if multiple {kib} instances connected to the same {es} cluster use different encryption keys. In this case, you might have a mix of saved objects encrypted with different keys, and every {kib} instance can only deal with a specific subset of objects. To fix this, you must choose a single primary encryption key for `xpack.encryptedSavedObjects.encryptionKey`, move all other encryption keys to `xpack.encryptedSavedObjects.keyRotation.decryptionOnlyKeys`, and sync this configuration across all {kib} instances.
```

============================================================================

> At some point you might want to dispose of old encryption keys completely. Make sure there are no saved objects that {kib} encrypted with these encryption keys. You can use the <<saved-objects-api-rotate-encryption-key, rotate encryption key API>> to determine which existing saved objects require decryption-only keys and re-encrypt them with the primary key.

Reviewer: ```suggestion
At some point you might want to dispose of old encryption keys completely. Make sure there are no saved objects that {kib} encrypted with these encryption keys. You can use the <<saved-objects-api-rotate-encryption-key, rotate encryption key API>> to determine which existing saved objects require decryption-only keys and re-encrypt them with the primary key.
```
```suggestion
At some point, you might want to dispose of old encryption keys completely. Make sure there are no saved objects that {kib} encrypted with these encryption keys. You can use the <<saved-objects-api-rotate-encryption-key, rotate encryption key API>> to determine which existing saved objects require decryption-only keys and re-encrypt them with the primary key.
```

> At some point you may want to dispose old encryption keys completely, but before that you should make sure there are no saved objects left that {kib} encrypted with the encryption key you're going to dispose. You can use <<saved-objects-api-rotate-encryption-key, Rotate encryption key API>> for that purpose. It will figure out which existing saved objects still require decryption-only keys and re-encrypt them with the primary key.

Reviewer: ```suggestion
At some point you might want to dispose of old encryption keys completely. Make sure there are no saved objects that {kib} encrypted with these encryption keys. You can use the <<saved-objects-api-rotate-encryption-key, rotate encryption key API>> to determine which existing saved objects require decryption-only keys and re-encrypt them with the primary key.
```

[role="xpack"]
[[xpack-security-secure-saved-objects]]

> === Secure Saved Objects

Author: __note:__ didn't want to be too specific and say `Encrypt`, used more generic `Secure` word....

> In {kib} you usually deal with dashboards, visualizations, alerts, actions, advanced settings, spaces etc. These are so called *Saved Objects* that {kib} stores in a dedicated internal {es} index. If such an object includes any sensitive information, for example a PagerDuty’s integration key or email server credentials used by the alert action, {kib} encrypts it and makes sure it cannot be accidentally leaked or tampered with. Moreover, {kib} records any access attempt to such sensitive information in the audit logs. 

Author: __note:__ That's the shortest introduction to `Saved Objects` I could come up with, happy to get better ideas! :see_no_evil:

> Having sensitive information encrypted means that even if a malicious party can somehow get access to the {kib} internal indices, for example through improperly secured backup, they won't be able to extract that information without knowing an encryption key usually stored separately in a keystore or a configuration file. That's why it's critical to use a cryptographically strong encryption key and keep it safe.

Author: Yeah, I think I'll just remove that sentence. We should either include that everywhere (including reporting and session keys) or don't mention at all. But I don't want to waste our time on that since I hope we will eventually migrate to a single admin-configurable master key and that one will have a proper doc section.

> Here is how your `kibana.yml` can look like if you define an encryption key:

Author: :+1: I copied this "reference to `kibana.yml`" and the one in `security-settings.asciidoc` from our existing sections, I think at some point we need to do a single proof-reading pass through our entire docs and make them look consistent :slightly_smiling_face:

> Here is how your `kibana.yml` can look like if you define an encryption key:

Reviewer 2: ++ that's a great idea

As a rule of thumb, when you change an encryption key you still keep the old one for some time. While {kib} will only use a new encryption key to encrypt all new and updated data it still may need to know the old one to decrypt data that was encrypted using that old key. Moreover, it's perfectly possible to have multiple old keys used only for decryption. It's worth noting that {kib} doesn't re-encrypt existing Saved Objects with the new encryption key automatically, re-encryption only happens when you update existing object or use bulk key rotation API explained later on.

Here is how your `kibana.yml` can look like if you want to use key rotation functionality:

[source,yaml]
--------------------------------------------------------------------------------
xpack.encryptedSavedObjects:
  encryptionKey: "min-32-byte-long-NEW-encryption-key" <1>
  keyRotation:
    decryptionOnlyKeys: ["min-32-byte-long-OLD#1-encryption-key", "min-32-byte-long-OLD#2-encryption-key"] <2>
--------------------------------------------------------------------------------

> <1> Encryption key {kib} will use to encrypt all new or updated Saved Objects, known as a primary encryption key.

Reviewer 2: ```suggestion
<1> Encryption key {kib} will use to encrypt all new or updated Saved Objects, known as the primary encryption key.
```

[NOTE]
============================================================================
You may also leverage this functionality if, by mistake, you ended up with multiple {kib} instances connected to the same {es} cluster that use different encryption keys. In this case you may have a mix of Saved Objects encrypted with different keys and every {kib} instance can deal only with a specific subset of objects. To fix this you'll need to choose a single primary encryption key for `xpack.encryptedSavedObjects.encryptionKey`, move all other encryption keys to `xpack.encryptedSavedObjects.keyRotation.decryptionOnlyKeys` and sync this configuration across all {kib} instances.
============================================================================

At some point you may want to dispose old encryption keys completely, but before that you should make sure there are no Saved Objects left that {kib} encrypted with the encryption key you're going to dispose. You can use bulk key rotation API for that purpose. It will figure out which existing Saved Objects still require decryption-only keys and re-encrypt them with the primary key. It may take a while if you have lots of such Saved Objects:

> 

Reviewer 2: What do you think about extracting this out to the [Saved Objects API docs](https://github.com/elastic/kibana/blob/1e3f14395c5e720c0e3a21370387e80d27b8b653/docs/api/saved-objects.asciidoc), and then link to it from here?

I think we should also update this to match the format of the other REST API docs ([example](https://github.com/elastic/kibana/blob/0b675b89084b18faa1db1ca99ecd500a78af8f57/docs/api/saved-objects/export.asciidoc))

Author: Yeah, that's a great idea! Will do.

It's also possible to rotate encryption key only for a specific type of the Saved Objects using an optional `type` query string parameter (default: all types), or/and specify a max batch size with an optional `batchSize` query string parameter (default and max: 10000):

[NOTE]
============================================================================
We optimized default parameters for speed and don't expect you to change any of those without a strong need. However, if you're experiencing any issues with this API you may want to decrease a batch size or rotate encryption key for the specific types only. In this case you may need to run key rotation multiple times in a row.
============================================================================

[source,sh]
--------------------------------------------------
$ curl -X POST /api/encrypted_saved_objects/_rotate_key?type=alert&batchSize=5000 <1>
--------------------------------------------------
// KIBANA

<1> {kib} will rotate encryption key only for the Saved Objects with the `alert` type in batches of 5000 objects.

> For the time being if you specify decryption-only keys {kib} may try to use them all if it cannot decrypt the Saved Object using the primary encryption key. In most of the cases this overhead is negligible, but if you're constantly dealing with the thousands of the Saved Objects that are encrypted and experiencing performance issues you may want to try to rotate encryption key for those objects.

Author: __note:__ I'm torn on this one, maybe it's more confusing than useful?

Reviewer 2: I feel like this would be better suited at the top of the API doc (once extracted, see other comment). This does a nice job of explaining why you might want to use this API, so I think it's worthwhile to keep in.

Author: Yep, moved, thanks!
